Support resource-based bootstrapping for backend.

fspmarshall opened the pull request:
```
Support resource-based bootstrapping for backend

Outside of static configuration, most of the persistent state of an
auth server exists as a collection of resources, stored in its
backend.  The resource API also forms the basis of Teleport's more
advanced dynamic configuration options.

This commit extends the usefulness of the resource API by adding
the ability to bootstrap backend state with a set of previously
exported resources.  This allows the resource API to serve as a
rudimentary backup/migration tool.

Notes: This features is a work in progress, and very easy to misuse;
while it will prevent you from overwriting the state of an existing
auth server, it won't stop you from bootstrapping into a wildly
misconfigured state.  In general, resource-based bootstrapping is
not a complete solution for backup or migration.
```
---
Opening this up for early feedback; some additional tweaks need to be put in place before this is very useful...

The main feature being added here is the ability to pass a collection of resources to `teleport start` via the `--bootstrap` option which will be used to pre-configure backend state on first start.  In the interest of usability, the `tctl get` subcommand has been extended to accept a comma-separated list of resource references to allow multiple resources or resource kinds to be exported at once.

This PR addresses most, but not all, of #2480 (in particular, it does not capture the config file).  I specifically opted for the `bootstrap` naming convention over the `backup` naming discussed there as I felt `backup` was potentially misleading (even with config file capture).

What works right now:

```bash
# export dynamic state from a local teleport server
tctl get all > state.yaml

# set up empty data dir & config for fresh server
mkdir fresh && cat > fresh.yaml << EOF
teleport:
  data_dir: fresh
EOF

# bootstrap fresh server (may need to kill the other)
teleport start --config fresh.yaml --bootstrap state.yaml

# from another terminal
tctl --config fresh.yaml get all
# your state here!
```

The catch:

The `tctl` utility cannot export user auth materials (e.g. two-factor tokens) yet, so while transferred users' current login credentials remain valid (as do credentials created with `tctl auth sign`), users are out of luck once their logins expire & they need to re-up.

Outstanding work:
- Add support two-factor tokens.
- Add support password hashes.
- Ensure that the advanced features like connectors & trusted clusters fully translate (mostly untested).

webvictim commented on lib/auth/init.go:
```diff
@@ -154,6 +158,23 @@ func Init(cfg InitConfig, opts ...AuthServerOption) (*AuthServer, error) {
+			return nil, trace.Wrap(err)
+		}
+		if firstStart {
+			log.Infof("Applying %v bootrstrap resources (first initialization)", len(cfg.Resources))
```
`bootstrap`

webvictim commented on tool/tctl/common/collection.go:
```diff
@@ -83,6 +91,13 @@ type namespaceCollection struct {
 
+func (c *namespaceCollection) resources() (r []services.Resource) {
+	for _, resource := range c.namespaces {
+		r = append(r, &resource)
```
Might be missing something but does this code deliberately get the address with `&resource` while all other instances of `append(r, resource)` in this file aren't?

fspmarshall replied:
Annoyingly, yes.  This is on purpose though; all the other resource types are interface types wrapping struct pointers, so they coerce to `services.Resource` directly.  The `services.Namespace` type is a concrete struct whose pointer coerces to `services.Resource`, so it has to be handled differently.

webvictim commented on lib/services/resource.go:
```diff
@@ -613,3 +787,43 @@ func (r *Ref) Set(v string) error {
+func ParseRefs(refs string) (Refs, error) {
+	isComma := func(r rune) bool { return r == ',' }
+	var parsed []Ref
+	split := strings.FieldsFunc(strings.TrimSpace(refs), isComma)
```
Do we currently make the explicit assertion elsewhere in Teleport that a comma can never be part of a resource name? Just trying to make sure we don't ever run into an issue where someone called their role "admins,1" which might break this.

fspmarshall replied:
Hmm, good catch.  I don't think we currently have any means to enforce a requirement that broad; different resources have different rules.  I suppose supporting backslash-escaping should cover the usability issue (e.g. `tctl get 'role/admins\,1'`), though in the future I it would be nice to globally enforce an allowed charset for resource names.

fspmarshall replied:
So it turns out we do have an enforcement point (kinda).  The backend restricts the charset of all keys, which indirectly restricts the charset of resource names.  Looks like commas are rejected.  Its not foolproof though; you can still create a resource with `/` in its name, which `tctl` will not be able to parse correctly.

webvictim reviewed (commented):
This looks quite reasonable to me - I'm still quite new to Go so afraid I'm not really able to get into the weeds with exactly how everything fits together, but I like the idea a lot and can see cases where it'd be very useful - not least of all in helping to build Teleport demo clusters which would benefit from having roles/trusted clusters/users adding straight away at startup and not relying on sidecar containers to do it.

klizhentas commented:
I think this is a good start, do you plan on adding a more generic `tctl dump/backup` command? that will take all the resources we support?

klizhentas commented:
Also, going through the code today, so will leave some comments.

klizhentas commented on lib/services/local/resource.go:
```diff
@@ -0,0 +1,365 @@
+//
+// NOTE: This function is non-atomic and performs no internal synchronization;
+// backend must be locked by caller when operating in parallel environment.
+func CreateResources(b backend.Backend, resources ...services.Resource) error {
```
can you add context as a first parameter here? this way you can really pass context from somewhere else instead of using `context.TODO()` below

klizhentas commented on lib/services/local/resource.go:
```diff
@@ -0,0 +1,365 @@
+// a given resource.  If `rsc` is not one of the supported resource types,
+// a `trace.NotImplementedError` is returned.
+func ItemizeResource(resource services.Resource) (*backend.Item, error) {
+	// If resource is of unknown type, attempt to cast it to a concrete
```
Instead of relying on Raw here, I think you should apply your own `TODO` item in the file-config package and convert to concrete types, so this part does not have a dependency on `Raw` that was supposed to be used during migrations only.

fspmarshall replied:
The `TODO` is fixed in upcoming push; some parts of `local/resource.go` will continue to rely on `services.UnknownResource` for now though (this is how `tctl` deals with unmarshaling mixed resource types as well).

klizhentas commented on lib/services/local/resource.go:
```diff
@@ -0,0 +1,365 @@
+// ItemizeResource attempts to construct an instance of `backend.Item` from
+// a given resource.  If `rsc` is not one of the supported resource types,
+// a `trace.NotImplementedError` is returned.
+func ItemizeResource(resource services.Resource) (*backend.Item, error) {
```
I would call it `ResourceToItem` instead of `Itemize`, both are correct, but I think we use the first notation more frequently throughout the codebase, similarly, `ResourceFromItem` is a mirror of that

fspmarshall replied:
Works for me; I was trying to follow the paradigm of  `Marshal<type>`/`Unmarshal<type>`, but `Deitemize` really is a mouthful lol.
How about `ItemFromResource` and `ItemToResource`?  I'd like to preserve the consistency of prefix for the different concrete functions if possible (e.g. `itemizeUser => itemFromUser`, `deitemizeCertAuthority => itemToCertAuthority`, etc).

klizhentas commented on lib/services/local/resource_test.go:
```diff
@@ -0,0 +1,183 @@
+
+func (r *ResourceSuite) dumpResources(c *check.C) []services.Resource {
+	startKey := []byte{0}
+	endKey := []byte{
```
could be helper function in the backend package `KeyRangeEnd()` or something like that to abstract away the magic constant

klizhentas commented on lib/services/local/resource_test.go:
```diff
@@ -0,0 +1,183 @@
+	s := NewIdentityService(r.bk)
+	b, err := s.GetUser("bob")
+	c.Assert(err, check.IsNil)
+	if !bob.Equals(b) {
```
could be a one liner: `c.Assert(bob.Equals(b), Equals, true, Commentf("dynamically inserted user does not match)`

klizhentas commented on lib/services/resource.go:
```diff
@@ -310,6 +310,180 @@ func SkipValidation() MarshalOption {
+}
+
+func init() {
+	// Register marshaler for user resources.
```
this comment just repeats the logic below, so it's redundant

klizhentas commented on lib/services/local/resource.go:
```diff
@@ -0,0 +1,365 @@
+	var err error
+	switch r := resource.(type) {
+	case services.User:
+		item, err = itemizeUser(r)
```
can you use the registered marshalers that you added here instead of another type switch?

fspmarshall replied:
I can, and I started to at one point, but then decided against it.  Basically, in order to attach the `Resource <=> Item` conversions to the marshalers, I'd need to make a decent number of organizational changes, including making a lot values which are currently private members of `lib/services/local` into public members of `lib/services`.

At the time, I decided this was counter to the prevailing organizational philosophy since the `lib/services/local` package keeps all the logic related to inserting resources into the backend heavily encapsulated.

What do you think is the right move here?

russjones commented on lib/config/configuration.go:
```diff
@@ -67,6 +69,8 @@ type CommandLineFlags struct {
 	AdvertiseIP string
 	// --config flag
 	ConfigFile string
+	// --bootstrap flag
```
```
// Bootstrap flag contains a YAML file that defines a set of resources to bootstrap
// a cluster.
```

russjones commented on lib/config/configuration.go:
```diff
@@ -118,6 +122,30 @@ func ReadConfigFile(cliConfigPath string) (*FileConfig, error) {
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	decoder := kyaml.NewYAMLOrJSONDecoder(reader, 32*1024)
```
Can you use `utils.ReadYAML` instead? Also can you make `32*1024` a constant?

fspmarshall replied:
I can definitely make `32*1024` a constant.  That value is used everywhere that `NewYAMLOrJSONDecoder` is called, where would be the appropriate place to put a commonly used constant like that?

In terms of using `utils.ReadYAML`, I'm happy to, but I'm not sure how lol (golang newbie problems).  From what I can tell, `utils.ReadYAML` dynamically creates everything as a bunch of primitive values (maps,strings,etc).  Does golang have some kind of mechanism for attempting to convert a composite of primitives to a concrete type?

fspmarshall committed:
Support resource-based bootstrapping for backend.

Outside of static configuration, most of the persistent state of an
auth server exists as a collection of resources, stored in its
backend.  The resource API also forms the basis of Teleport's more
advanced dynamic configuration options.

This commit extends the usefulness of the resource API by adding
the ability to bootstrap backend state with a set of previously
exported resources.  This allows the resource API to serve as a
rudimentary backup/migration tool.

Notes: This features is a work in progress, and very easy to misuse;
while it will prevent you from overwriting the state of an existing
auth server, it won't stop you from bootstrapping into a wildly
misconfigured state.  In general, resource-based bootstrapping is
not a complete solution for backup or migration.

fspmarshall committed:
update e-ref

kontsevoy commented:
Neat. My biggest UX gripe is how verbose it gets, i.e. the most natural and probably the most common use case is to export everything with secrets, so why then we require users to pass so many arguments for the most common use case?

Usually, you'd want to the most common action to be done WITHOUT any arguments at all, only passing flags when you want to deviate from the default.
